### ISSUE TITLE
Wired issus while subclassing MoyaProvider

### ISSUE THREAD

**Issue by iShawnWang** [Moya/Moya#982]
Moya i am Using
```
PODS:
  - Alamofire (4.3.0)
  - Moya (8.0.2):
    - Moya/Core (= 8.0.2)
  - Moya/Core (8.0.2):
    - Alamofire (~> 4.1)
    - Result (~> 3.0)
  - Result (3.1.0)
```

---

I try to subclass MoyaProvider like [ReactiveSwiftMoyaProvider.swift](https://github.com/Moya/Moya/blob/master/Sources/ReactiveMoya/ReactiveSwiftMoyaProvider.swift) , so that I can set a global http header for all my requests

## BaseProvider.swift

```swift
import UIKit
import Moya

open class BaseProvider<Target>: MoyaProvider<Target> where Target: TargetType {
    override open func endpoint(_ token: Target) -> Endpoint<Target> {
        let endpoint = endpointClosure(token)
        return endpoint.adding(newHTTPHeaderFields: ["golbalHeader":"h"])
    }
}
```

## Router.swift

```swift
import UIKit
import Moya

enum LessonRouter: String,TargetType{
    case lessonList = ""
    
    var baseURL: URL {
        return URL(string:"")!
    }
        
    var path: String{
        switch self {
        case .lessonList:
            return self.rawValue
        }
    }

    var method: Moya.Method {
        return .get
    }
    
    var parameters: [String: Any]? {
        return nil
    }
    
    var parameterEncoding: ParameterEncoding {
        return URLEncoding.methodDependent
    }
    
    var sampleData: Data {
        return Data()
    }
    
    var task: Task {
        return .request
    }
    
    var validate: Bool {
        return false
    }
    
}
```

## Finally In my ViewController

```swift
import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
//        let lessonProvider = BaseProvider<LessonRouter>()
    }
}

```
if I uncomment that line, I got following error

```
... 
... a lot useless msg here
... 

1.	While emitting SIL for 'viewDidLoad' at /Users/pi/workSpace/MoyaDemo/MoyaDemo/ViewController.swift:13:14
2.	While emitting reabstraction thunk in SIL function @_TTRXFo_ix_oGC4Moya8Endpointx__XFo_dO8MoyaDemo12LessonRouter_oGS0_S2___

```

What the hell with my code here, or just anyone poisoning my Xcode !!!
Help plz  ~

# :(

**Comment by iShawnWang** [comment]
I just create a whole new project to reproduce this issue, [here it is](https://github.com/iShawnWang/MoyaDemo), feel free to run to see this problem

**Comment by scottrhoyt** [comment]
@iShawnWang thanks for reporting this and creating a repro project. I can confirm that I experienced the same compiler crash from that project. I haven't had a chance to look into it much further, however, because these sorts of compiler crashes can take a while to debug 😞 . In the past when I have experienced this it has been related to issues with generics (many times having to do with generics and inheritance). I don't know if that is the issue here or not though, but it's my first guess. Have you had a chance to see if this behavior is the same when compiling with Swift 3.1?
